File: src/runtime/memory/goo_region_allocator.c
/**
 * goo_region_allocator.c
 * 
 * Implementation of region-based memory allocator for the Goo programming language.
 * Region allocators allow for efficient scoped memory management.
 */

#include <stdlib.h>
#include <string.h>
#include <assert.h>
#include <stdio.h>
#include "../memory/goo_allocator.h"

// Region allocation block
typedef struct RegionBlock {
    struct RegionBlock* next;     // Next block in region
    size_t size;                  // Size of this block
    size_t used;                  // Bytes used in this block
    unsigned int region_depth;    // Region depth this block belongs to
    char data[];                  // Flexible array member for actual data
} RegionBlock;

// Region tracking
typedef struct RegionInfo {
    struct RegionInfo* parent;    // Parent region
    RegionBlock* blocks;          // Blocks allocated in this region
    unsigned int depth;           // Nesting depth
} RegionInfo;

// Default block size for region allocations
#define REGION_DEFAULT_BLOCK_SIZE (64 * 1024)  // 64K

// Minimum block size
#define REGION_MIN_BLOCK_SIZE 1024  // 1K

// Helper to calculate required block size
static size_t region_calc_block_size(size_t requested_size) {
    // Ensure we have at least REGION_MIN_BLOCK_SIZE bytes
    size_t block_size = REGION_DEFAULT_BLOCK_SIZE;
    
    // If requested size is larger than default, allocate a custom-sized block
    if (requested_size > REGION_DEFAULT_BLOCK_SIZE - sizeof(RegionBlock)) {
        block_size = requested_size + sizeof(RegionBlock);
        // Round up to multiple of page size (assume 4K)
        block_size = (block_size + 4095) & ~4095;
    }
    
    return block_size;
}

// Create a new region block
static RegionBlock* region_block_create(GooAllocator* parent, size_t size, unsigned int depth) {
    size_t block_size = region_calc_block_size(size);
    RegionBlock* block = parent->alloc(parent, block_size, 16, GOO_ALLOC_ZERO);
    if (!block) return NULL;
    
    block->next = NULL;
    block->size = block_size - sizeof(RegionBlock);
    block->used = 0;
    block->region_depth = depth;
    
    return block;
}

// Allocate memory from a region block
static void* region_block_alloc(RegionBlock* block, size_t size, size_t alignment) {
    // Calculate alignment padding needed
    size_t offset = block->used;
    size_t padding = ((offset + alignment - 1) & ~(alignment - 1)) - offset;
    size_t total_size = size + padding;
    
    // Check if there's enough space
    if (block->used + total_size > block->size) {
        return NULL;
    }
    
    // Allocate from the block
    void* ptr = &block->data[offset + padding];
    block->used += total_size;
    
    return ptr;
}

// Region allocator functions
static void* region_alloc(GooAllocator* self, size_t size, size_t alignment, GooAllocOptions options) {
    GooRegionAllocator* region = (GooRegionAllocator*)self;
    if (!region || !region->parent) return NULL;
    
    // Update statistics if enabled
    if (self->track_stats) {
        self->stats.total_allocations++;
    }
    
    // Find the current region info
    RegionInfo* current_region = (RegionInfo*)region->regions;
    if (!current_region) {
        // No active region, use parent allocator directly
        void* ptr = region->parent->alloc(region->parent, size, alignment, options);
        
        if (self->track_stats && ptr) {
            self->stats.bytes_allocated += size;
            self->stats.allocation_count++;
            if (self->stats.bytes_allocated > self->stats.max_bytes_allocated) {
                self->stats.max_bytes_allocated = self->stats.bytes_allocated;
            }
        }
        
        return ptr;
    }
    
    // Try to allocate from the current region's first block
    RegionBlock* block = current_region->blocks;
    if (block) {
        void* ptr = region_block_alloc(block, size, alignment);
        if (ptr) {
            if (self->track_stats) {
                self->stats.bytes_allocated += size;
                self->stats.allocation_count++;
                if (self->stats.bytes_allocated > self->stats.max_bytes_allocated) {
                    self->stats.max_bytes_allocated = self->stats.bytes_allocated;
                }
            }
            
            // Zero memory if requested
            if (options & GOO_ALLOC_ZERO) {
                memset(ptr, 0, size);
            }
            
            return ptr;
        }
    }
    
    // Couldn't allocate from existing block, create a new one
    size_t block_size = region_calc_block_size(size);
    RegionBlock* new_block = region_block_create(region->parent, size, current_region->depth);
    if (!new_block) {
        if (self->track_stats) {
            self->stats.failed_allocations++;
        }
        
        // Handle allocation failure based on strategy
        switch (self->strategy) {
            case GOO_ALLOC_STRATEGY_PANIC:
                fprintf(stderr, "FATAL: Out of memory in region allocator\n");
                abort();
                break;
                
            case GOO_ALLOC_STRATEGY_RETRY:
                if (self->out_of_mem_fn) {
                    self->out_of_mem_fn();
                    // Try again
                    return region_alloc(self, size, alignment, options);
                }
                // Fall through to NULL if no handler
                
            case GOO_ALLOC_STRATEGY_NULL:
            default:
                return NULL;
        }
    }
    
    // Add new block to the region
    new_block->next = current_region->blocks;
    current_region->blocks = new_block;
    
    // Allocate from the new block
    void* ptr = region_block_alloc(new_block, size, alignment);
    if (!ptr) {
        // This should never happen since we just created a block big enough
        if (self->track_stats) {
            self->stats.failed_allocations++;
        }
        return NULL;
    }
    
    if (self->track_stats) {
        self->stats.bytes_allocated += size;
        self->stats.bytes_reserved += block_size;
        self->stats.allocation_count++;
        if (self->stats.bytes_allocated > self->stats.max_bytes_allocated) {
            self->stats.max_bytes_allocated = self->stats.bytes_allocated;
        }
    }
    
    // Zero memory if requested
    if (options & GOO_ALLOC_ZERO) {
        memset(ptr, 0, size);
    }
    
    return ptr;
}

// Reallocation in a region allocator (note: only works for the most recent allocation)
static void* region_realloc(GooAllocator* self, void* ptr, size_t old_size, size_t new_size, 
                            size_t alignment, GooAllocOptions options) {
    GooRegionAllocator* region = (GooRegionAllocator*)self;
    if (!region || !region->parent) return NULL;
    
    // If ptr is NULL, this is just an allocation
    if (!ptr) {
        return region_alloc(self, new_size, alignment, options);
    }
    
    // If new_size is 0, this is just a free (which is a no-op in regions)
    if (new_size == 0) {
        return NULL;
    }
    
    // Find the current region info
    RegionInfo* current_region = (RegionInfo*)region->regions;
    if (!current_region) {
        // No active region, use parent allocator directly
        return region->parent->realloc(region->parent, ptr, old_size, new_size, alignment, options);
    }
    
    // Check if this is the most recent allocation in the current block
    RegionBlock* block = current_region->blocks;
    if (block) {
        char* data_end = &block->data[block->used];
        char* ptr_end = (char*)ptr + old_size;
        
        // If this pointer is at the end of the block's used memory, we can just adjust the block size
        if (ptr_end == data_end) {
            // Check if we can expand in place
            if (new_size <= old_size + (block->size - block->used)) {
                // Can expand in place
                block->used += (new_size - old_size);
                
                if (self->track_stats) {
                    self->stats.bytes_allocated += (new_size - old_size);
                    if (self->stats.bytes_allocated > self->stats.max_bytes_allocated) {
                        self->stats.max_bytes_allocated = self->stats.bytes_allocated;
                    }
                }
                
                // Zero any new memory if requested
                if (options & GOO_ALLOC_ZERO && new_size > old_size) {
                    memset((char*)ptr + old_size, 0, new_size - old_size);
                }
                
                return ptr;
            }
        }
    }
    
    // Can't realloc in place, allocate new memory and copy
    void* new_ptr = region_alloc(self, new_size, alignment, options & ~GOO_ALLOC_ZERO);
    if (!new_ptr) return NULL;
    
    // Copy old data
    memcpy(new_ptr, ptr, old_size < new_size ? old_size : new_size);
    
    // Zero any new memory if requested
    if (options & GOO_ALLOC_ZERO && new_size > old_size) {
        memset((char*)new_ptr + old_size, 0, new_size - old_size);
    }
    
    // Note: We don't free the old memory in a region allocator
    
    return new_ptr;
}

// Free in a region allocator (no-op, memory is freed when the region is ended)
static void region_free(GooAllocator* self, void* ptr, size_t size, size_t alignment) {
    // No individual free in region allocator, memory is freed when region is ended
    GooRegionAllocator* region = (GooRegionAllocator*)self;
    if (!region) return;
    
    if (self->track_stats) {
        self->stats.total_frees++;
        self->stats.bytes_allocated -= size;
        self->stats.allocation_count--;
    }
}

// Destroy the region allocator
static void region_destroy(GooAllocator* self) {
    GooRegionAllocator* region = (GooRegionAllocator*)self;
    if (!region || !region->parent) return;
    
    // Free all regions and their blocks
    RegionInfo* info = (RegionInfo*)region->regions;
    while (info) {
        RegionInfo* parent = info->parent;
        
        // Free all blocks in this region
        RegionBlock* block = info->blocks;
        while (block) {
            RegionBlock* next = block->next;
            region->parent->free(region->parent, block, block->size + sizeof(RegionBlock), 16);
            block = next;
        }
        
        // Free the region info
        region->parent->free(region->parent, info, sizeof(RegionInfo), 16);
        
        info = parent;
    }
    
    // Free the region allocator itself
    region->parent->free(region->parent, region, sizeof(GooRegionAllocator), 16);
}

// Create a region allocator
GooRegionAllocator* goo_region_allocator_create(GooAllocator* parent) {
    if (!parent) return NULL;
    
    GooRegionAllocator* region = parent->alloc(parent, sizeof(GooRegionAllocator), 16, GOO_ALLOC_ZERO);
    if (!region) return NULL;
    
    // Initialize allocator interface
    region->allocator.alloc = region_alloc;
    region->allocator.realloc = region_realloc;
    region->allocator.free = region_free;
    region->allocator.destroy = region_destroy;
    region->allocator.strategy = GOO_ALLOC_STRATEGY_NULL;
    region->allocator.out_of_mem_fn = NULL;
    region->allocator.context = region;
    region->allocator.track_stats = true;
    memset(&region->allocator.stats, 0, sizeof(GooAllocStats));
    
    // Initialize region-specific data
    region->parent = parent;
    region->regions = NULL;
    region->current_depth = 0;
    
    return region;
}

// Begin a new memory region
void goo_region_begin(GooRegionAllocator* region) {
    if (!region || !region->parent) return;
    
    // Create a new region info
    RegionInfo* info = region->parent->alloc(region->parent, sizeof(RegionInfo), 16, GOO_ALLOC_ZERO);
    if (!info) return;
    
    // Initialize the region info
    info->parent = (RegionInfo*)region->regions;
    info->blocks = NULL;
    info->depth = ++region->current_depth;
    
    // Set as current region
    region->regions = info;
}

// End the current memory region
void goo_region_end(GooRegionAllocator* region) {
    if (!region || !region->parent) return;
    
    RegionInfo* info = (RegionInfo*)region->regions;
    if (!info) return;
    
    // Update region depth
    region->current_depth--;
    
    // Update the current region to the parent
    region->regions = info->parent;
    
    // Free all blocks in this region
    RegionBlock* block = info->blocks;
    while (block) {
        RegionBlock* next = block->next;
        
        // Update stats
        if (region->allocator.track_stats) {
            region->allocator.stats.bytes_reserved -= (block->size + sizeof(RegionBlock));
        }
        
        region->parent->free(region->parent, block, block->size + sizeof(RegionBlock), 16);
        block = next;
    }
    
    // Free the region info
    region->parent->free(region->parent, info, sizeof(RegionInfo), 16);
} 

File: src/runtime/memory/goo_region_allocator.zig
const std = @import("std");
const c = @cImport({
    @cInclude("stdlib.h");
    @cInclude("string.h");
});

const allocator_core = @import("goo_allocator_core.zig");
const Allocator = allocator_core.Allocator;
const AllocOptions = allocator_core.AllocOptions;
const DEFAULT_ALIGNMENT = allocator_core.DEFAULT_ALIGNMENT;

// Constants
const DEFAULT_REGION_SIZE = 1024 * 1024; // 1 MB
const MIN_REGION_SIZE = 64 * 1024; // 64 KB
const MAX_ALLOCATION_SIZE = 1024 * 1024; // Maximum size for a single allocation

// Region structure
const Region = struct {
    next: ?*Region, // Linked list of regions
    size: usize, // Total size of this region
    used: usize, // Used space in this region
    data: [*]u8, // Memory area

    // Create a new region
    fn create(size: usize) ?*Region {
        // Allocate the region header and data in one block
        const header_size = @sizeOf(Region);
        const total_size = header_size + size;

        const ptr = c.malloc(total_size);
        if (ptr == null) return null;

        // Initialize the region
        const region = @as(*Region, @ptrCast(@alignCast(ptr)));
        region.* = .{
            .next = null,
            .size = size,
            .used = 0,
            .data = @as([*]u8, @ptrCast(@alignCast(ptr))) + header_size,
        };

        return region;
    }

    // Destroy a region
    fn destroy(region: *Region) void {
        c.free(region);
    }

    // Allocate memory from this region
    fn alloc(self: *Region, size: usize, alignment: usize) ?*anyopaque {
        // Ensure alignment
        const current_ptr = @intFromPtr(&self.data[self.used]);
        const aligned_ptr = (current_ptr + alignment - 1) & ~(alignment - 1);
        const padding = aligned_ptr - current_ptr;

        // Check if we have enough space
        const total_needed = padding + size;
        if (self.used + total_needed > self.size) {
            return null;
        }

        // Update used count and return aligned pointer
        self.used += total_needed;
        return @as(*anyopaque, @ptrFromInt(aligned_ptr));
    }

    // Reset a region for reuse
    fn reset(self: *Region) void {
        self.used = 0;
    }
};

// Allocation header for tracking
const AllocHeader = struct {
    size: usize, // Size of this allocation
    payload: [*]u8, // Pointer to the actual data
};

// Region allocator implementation
pub const RegionAllocator = struct {
    // Base allocator interface
    allocator: Allocator,

    // Region-specific state
    regions: ?*Region,
    current_region: ?*Region,
    region_size: usize,
    allow_large_allocations: bool,

    // Initialize a region allocator
    pub fn init(region_size: usize, allow_large_allocations: bool) ?*RegionAllocator {
        // Validate parameters
        const actual_region_size = if (region_size < MIN_REGION_SIZE)
            DEFAULT_REGION_SIZE
        else
            region_size;

        // Allocate the allocator struct
        const self = @as(*RegionAllocator, @ptrCast(@alignCast(c.malloc(@sizeOf(RegionAllocator)) orelse return null)));

        // Initialize fields
        self.* = .{
            .allocator = .{
                .vtable = &regionVTable,
                .strategy = .null,
                .out_of_mem_fn = null,
                .context = self,
                .track_stats = true,
                .stats = .{},
            },
            .regions = null,
            .current_region = null,
            .region_size = actual_region_size,
            .allow_large_allocations = allow_large_allocations,
        };

        // Allocate initial region
        if (!self.addRegion(actual_region_size)) {
            c.free(self);
            return null;
        }

        return self;
    }

    // Add a new region
    fn addRegion(self: *RegionAllocator, min_size: usize) bool {
        // Determine region size (at least min_size, but normally the default size)
        const size = if (min_size > self.region_size) min_size else self.region_size;

        // Create the region
        const region = Region.create(size) orelse return false;

        // Add to the region list
        if (self.regions == null) {
            self.regions = region;
        } else if (self.current_region) |current| {
            current.next = region;
        }

        self.current_region = region;

        // Update stats
        if (self.allocator.track_stats) {
            self.allocator.stats.bytes_reserved += size;
        }

        return true;
    }

    // Get the allocation header to retrieve the original size
    fn getHeader(ptr: ?*anyopaque) ?*AllocHeader {
        if (ptr == null) return null;
        const header = @as(*AllocHeader, @ptrCast(@alignCast(ptr.?)));
        return header;
    }

    // Allocate memory from the region
    fn alloc(allocator: *Allocator, size: usize, alignment: usize, ret_addr: c_int) ?*anyopaque {
        if (size == 0) return null;

        const self = @as(*RegionAllocator, @ptrCast(@alignCast(allocator.context.?)));
        const actual_alignment = if (alignment == 0) DEFAULT_ALIGNMENT else alignment;

        // Calculate aligned sizes for the header
        const header_size = @sizeOf(AllocHeader);
        const header_aligned_size = std.mem.alignForward(usize, header_size, actual_alignment);
        const total_size = header_aligned_size + size;

        // Check for large allocations
        if (total_size > self.region_size) {
            // For very large allocations, we can either fail or create a custom-sized region
            if (!self.allow_large_allocations) {
                if (allocator.track_stats) {
                    allocator.stats.failed_allocations += 1;
                }
                return null;
            }

            // Create a custom-sized region
            if (!self.addRegion(total_size)) {
                if (allocator.track_stats) {
                    allocator.stats.failed_allocations += 1;
                }
                return null;
            }

            // The new region is now the current one
        }

        // Try to allocate from current region
        var ptr: ?*anyopaque = null;
        if (self.current_region) |current| {
            ptr = current.alloc(total_size, actual_alignment);
        }

        // If current region is full, allocate a new one
        if (ptr == null) {
            if (!self.addRegion(total_size)) {
                if (allocator.track_stats) {
                    allocator.stats.failed_allocations += 1;
                }
                return null;
            }

            // Try again with the new region
            ptr = self.current_region.?.alloc(total_size, actual_alignment);
            if (ptr == null) {
                // This should never happen since we just created a region of sufficient size
                if (allocator.track_stats) {
                    allocator.stats.failed_allocations += 1;
                }
                return null;
            }
        }

        // Set up the allocation header
        const header = @as(*AllocHeader, @ptrCast(@alignCast(ptr.?)));
        header.size = size;
        header.payload = @as([*]u8, @ptrCast(ptr.?)) + header_aligned_size;

        // Update stats
        if (allocator.track_stats) {
            allocator.stats.bytes_allocated += size;
            allocator.stats.allocation_count += 1;
            allocator.stats.total_allocations += 1;

            if (allocator.stats.bytes_allocated > allocator.stats.max_bytes_allocated) {
                allocator.stats.max_bytes_allocated = allocator.stats.bytes_allocated;
            }
        }

        // Zero memory if requested
        const alloc_options = AllocOptions.fromC(ret_addr);
        if (alloc_options.zero) {
            @memset(header.payload[0..size], 0);
        }

        return @as(*anyopaque, @ptrCast(header.payload));
    }

    // Realloc is similar to arena - we always allocate new memory
    fn realloc(allocator: *Allocator, ptr: ?*anyopaque, old_size: usize, new_size: usize, alignment: usize, ret_addr: c_int) ?*anyopaque {
        // Special cases
        if (ptr == null) {
            return RegionAllocator.alloc(allocator, new_size, alignment, ret_addr);
        }

        if (new_size == 0) {
            RegionAllocator.free(allocator, ptr, old_size, alignment);
            return null;
        }

        // Allocate new memory
        const new_ptr = RegionAllocator.alloc(allocator, new_size, alignment, ret_addr);
        if (new_ptr == null) {
            return null;
        }

        // Copy data from old to new
        const copy_size = @min(old_size, new_size);
        @memcpy(@as([*]u8, @ptrCast(new_ptr))[0..copy_size], @as([*]u8, @ptrCast(ptr))[0..copy_size]);

        // Update stats for the old allocation
        if (allocator.track_stats) {
            allocator.stats.bytes_allocated -|= old_size;
            allocator.stats.allocation_count -|= 1;
            allocator.stats.total_frees += 1;
        }

        return new_ptr;
    }

    // Individual frees track stats but don't actually free memory
    fn free(allocator: *Allocator, ptr: ?*anyopaque, size: usize, alignment: usize) void {
        _ = alignment;
        if (ptr == null) return;

        // Update stats
        if (allocator.track_stats) {
            // For more accurate stats, we can get the actual size from the header
            // But we can also just use the provided size for simplicity
            allocator.stats.bytes_allocated -|= size;
            allocator.stats.allocation_count -|= 1;
            allocator.stats.total_frees += 1;
        }
    }

    // Reset all regions for reuse
    pub fn reset(self: *RegionAllocator) void {
        var region = self.regions;
        while (region) |r| {
            r.reset();
            region = r.next;
        }

        // Reset the current region to the first one
        self.current_region = self.regions;

        // Reset stats
        if (self.allocator.track_stats) {
            self.allocator.stats.bytes_allocated = 0;
            self.allocator.stats.allocation_count = 0;
        }
    }

    // Destroy the allocator and all its regions
    fn destroy(allocator: *Allocator) void {
        const self = @as(*RegionAllocator, @ptrCast(@alignCast(allocator.context.?)));

        // Free all regions
        var region = self.regions;
        while (region != null) {
            const next = region.?.next;
            Region.destroy(region.?);
            region = next;
        }

        // Free the allocator itself
        c.free(self);
    }
};

// Allocator vtable for region allocator
const regionVTable = allocator_core.AllocatorVTable{
    .allocFn = RegionAllocator.alloc,
    .reallocFn = RegionAllocator.realloc,
    .freeFn = RegionAllocator.free,
    .destroyFn = RegionAllocator.destroy,
};

// Exported C API functions
export fn goo_region_allocator_create(region_size: usize, allow_large_allocations: bool) ?*Allocator {
    if (RegionAllocator.init(region_size, allow_large_allocations)) |region| {
        return &region.allocator;
    }

    return null;
}

export fn goo_region_allocator_reset(allocator: ?*Allocator) void {
    if (allocator == null) return;

    const self = @as(*RegionAllocator, @ptrCast(@alignCast(allocator.?.context.?)));
    self.reset();
}

export fn goo_region_allocator_destroy(allocator: ?*Allocator) void {
    if (allocator == null) return;

    allocator.?.destroy();
}
